lib/spack/spack/config.py

```diff
@@ -498,6 +499,9 @@ def update_config(self, section, update_data, scope=None, force=False):
             scope (str): scope to be updated
             force (str): force the update
         """
+        # Invalidate the cache
+        self._get_config_memoized.cache.clear()
```

Comment:
@scheibelp I share your concern. This is a lot faster than before. @haampie reported his use case going:
- From 11min. before #18482 
- To 90 min. with #18482
- To 4 min. with this PR

Unit tests are running 10-15 mins. faster than other recent PRs and I see speedups of 4x on simple `spack spec ...` commands. I'm already trying to intercept writes to invalidate the cache. I agree though that if it works, it would feel like taking a very fast shortcut through a minefield...

Reply:
> I think v3 should reflect the changes to v2

How does the user change v2 in the configuration? If the user goes through the API it should work if we clear the cache on commands that mutate the state of the config.

Reply:
> I'm already trying to intercept writes to invalidate the cache

I'm curious what your approach is.

It's possible things could be simplified by adding a `read_only` option to `get_config`. Most parts of the code retrieve the configuration to read it and in that case a `deepcopy` should not be required. I think I'd rather trust these parts of the code for the time being and add an issue to do the interception you suggest (which I think would address both the modification and speed concerns).

Reply:
> It's possible things could be simplified by adding a read_only option to get_config

This is not without drawbacks as it would make the API signature harder to understand. Can you clarify how a user would mutate v2 in your example above? I'm trying to understand if it is expected to go through `set_config` or other methods of the `Configuration` class.

Reply:
>  Can you clarify how a user would mutate v2 in your example above?

Config objects behave like nested dictionaries, so I would think things like

```
v2['foo'] = 'bar'
```

would be fair game. That being said I'm wondering if we generally wrap up all our modifications with `update_config` or `config.set` - invalidating the cache in those methods might be sufficient. That would require us to agree on updating the API and documentation for `Config` to reflect this requirement.

Reply:
> it would make the API signature harder to understand

I should also say that adding a cache that has to be cleared in disperse places (i.e. the locations you have added `_get_config_memoized_cache.clear()`) also complicates the API. I'd rather avoid that if possible. 

That could be simplified in your approach here by moving those calls into the `config` module. For example: never set `spack.config.config` but rather call a function like `spack.config.set_config`; `config.remove_scope` could also clear the cache automatically.

Reply:
> That being said I'm wondering [ ... ] That would require us to agree on updating the API and documentation for Config to reflect this requirement.

That's what I was thinking too. If we agree to go through those methods we have a much smaller surface to check. I think it's safe to say that unit tests are surely going through `Configuration` methods.

Reply:
> Adding a cache that has to be cleared in disperse places (i.e. the locations you have added _get_config_memoized_cache.clear()) also complicates the API. I'd rather avoid that if possible.

Ugh, I didn't get a refresh automatically, so I saw your comment just now. I did something that goes in the same direction with decorators in 27e0f32